core/trino-main/src/main/java/io/trino/FeaturesConfig.java

```diff
@@ -81,6 +83,7 @@ public class FeaturesConfig
      * default value is overwritten for fault tolerant execution in {@link #applyFaultTolerantExecutionDefaults()}}
      */
     private boolean exchangeCompressionEnabled;
+    private CompressionCodec exchangeCompressionCodec = LZ4;
```

Comment:
I ran TPCDS sf1000 benchmark on unpartitioned hive tables with ZSTD exchange compression in pipelined execution mode. The setup is a 6 node cluster of r6g.8xlarge machines on AWS.

[TPCDS unpartitioned parquet zstd exchange.pdf](https://github.com/trinodb/trino/files/13839896/TPCDS.unpartitioned.parquet.zstd.exchange.pdf)

<img width="924" alt="Screenshot 2024-01-05 at 2 18 59 PM" src="https://github.com/trinodb/trino/assets/4344846/7294c762-3489-4a91-bf20-de7e98613b5f">

Even though there is a big reduction in internal network data exchanged, there are still regressions in latency due to the high CPU cost of ZSTD compression.
I think this trade-off of more aggressive compression at high CPU cost doesn't make much sense for pipelined execution mode where we exchange data over internal network directly.
This trade off might make sense for filesystem based exchange in fault tolerant execution mode, but that also needs to be assessed with a benchmark.
@losipiuk can we do these changes in way where this config becomes specific to filesystem exchange in FTE ?

Reply:
Benchmark results with FTE
[FTE unpartitioned hive parquet sf1k zstd compression.pdf](https://github.com/trinodb/trino/files/13870656/FTE.unpartitioned.hive.parquet.sf1k.zstd.compression.pdf)

<img width="1358" alt="Screenshot 2024-01-09 at 1 26 54 PM" src="https://github.com/trinodb/trino/assets/4344846/92f44a03-3937-4329-afdf-3844b6cd5a21">

FTE is using LZ4 exchange compression by default already. Switching to ZSTD improves some queries but it's significantly worse overall. So I'm not seeing a compelling reason to add this option even for FTE.

Reply:
> Benchmark results with FTE [FTE unpartitioned hive parquet sf1k zstd compression.pdf](https://github.com/trinodb/trino/files/13870656/FTE.unpartitioned.hive.parquet.sf1k.zstd.compression.pdf)
> 
> <img alt="Screenshot 2024-01-09 at 1 26 54 PM" width="1358" src="https://private-user-images.githubusercontent.com/4344846/295137308-92f44a03-3937-4329-afdf-3844b6cd5a21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ3OTg3NzUsIm5iZiI6MTcwNDc5ODQ3NSwicGF0aCI6Ii80MzQ0ODQ2LzI5NTEzNzMwOC05MmY0NGEwMy0zOTM3LTQzMjktYWZkZi0zODQ0YjZjZDVhMjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDlUMTEwNzU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDQ5N2U3MGMzNDc3NTNhMTVjNzMzODZjNzU2YThlOTUwODcyZWRkNWU2M2NjMzBhYzgyOWNmMGI1NTU1MjhmNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.3PAyEPNHQzIjcY8Jqf1LTD_27ITvoIS30hjP8ETqelE">
> FTE is using LZ4 exchange compression by default already. Switching to ZSTD improves some queries but it's significantly worse overall. So I'm not seeing a compelling reason to add this option even for FTE.

You are right, according to the benchmark results, ZSTD should only be used in clusters with little network bandwidth.

We found that some gaming analytics customers of our company have clusters with strong CPU and poor network cards, maybe in this condition we can use it~  

I saw that `Huawei OpenLookeng`(originated from Trino 350) uses ZSTD in exchanges, but it was for their datacenter connector:  
![image](https://github.com/trinodb/trino/assets/26461591/6879daae-efe8-4a6e-930e-150af2e0bba3)